src/webapi/form_data.rs

```diff
@@ -0,0 +1,291 @@
+use webapi::blob::IBlob;
+use webapi::dom_exception::InvalidStateError;
+use webapi::element::IElement;
+use webapi::error::TypeError;
+use webapi::file::File;
+use webcore::serialization::JsSerialize;
+use webcore::try_from::TryFrom;
+use webcore::try_from::TryInto;
+use webcore::value::ConversionError;
+use webcore::value::Reference;
+use webcore::value::Value;
+
+/// The `FormData` interface provides a way to easily construct a set of key/value pairs
+/// representing form fields and their values, which can then be easily sent using the
+/// `XMLHttpRequest.send()` method. It uses the same format a form would use if the encoding type
+/// were set to `"multipart/form-data"`.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/FormData)
+// https://xhr.spec.whatwg.org/#formdata
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "FormData")]
+pub struct FormData( Reference );
+
+/// Represents a type of data stores in FormData.
+#[derive(Clone, Debug, PartialEq)]
+pub enum FormDataEntry {
+    /// File data
+    File( File ),
+    /// Text data
+    String( String )
+}
+
+error_enum_boilerplate! {
+    FormDataFromElementError,
+    InvalidStateError, TypeError
+}
+
+impl TryFrom< Value > for FormDataEntry {
+    type Error = ConversionError;
+
+    fn try_from(value: Value) -> Result< Self, Self::Error > {
+        let entry = match value {
+            Value::String(s) => FormDataEntry::String(s),
+            Value::Reference(r) => FormDataEntry::File(File(r)),
+            _ => return Err(ConversionError::type_mismatch(&value, "string or reference".into())),
+        };
+
+        Ok(entry)
+    }
+}
+
+impl TryFrom< Value > for Option< FormDataEntry > {
+    type Error = ConversionError;
+
+    fn try_from(value: Value) -> Result< Self, Self::Error > {
+        let entry = match value {
+            Value::Null => None,
+            Value::String(s) => Some(FormDataEntry::String(s)),
+            Value::Reference(r) => Some(FormDataEntry::File(File(r))),
+            _ => return Err(ConversionError::type_mismatch(&value, "null, string or reference".into())),
+        };
+
+        Ok(entry)
+    }
+}
+
+impl FormData {
+    /// Creates a new `FormData`.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData)
+    pub fn new() -> Self {
+        js! (
+            return new FormData();
+        ).try_into().unwrap()
+    }
+
+    /// Creates a new `FormData` from a form element.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData)
+    pub fn from_element<T>( form: &T ) -> Result< Self, FormDataFromElementError >
+        where T: IElement + JsSerialize {
```

Comment:
To be consistent with the rest of the library remove the `JsSerialize` bound and use `@{form.as_ref()}` instead of `@{form}`. (Sorry, it slipped my mind the first time I looked through this.) Same for other methods here.